cli/loader_lib.c

```diff
@@ -153,6 +154,82 @@ JL_DLLEXPORT const char * jl_get_libdir()
     return lib_dir;
 }
 
+#ifdef LIBSTDCXXPROBE
+// Write the path to the libstdcxx to load to "path", which should be a buffer of at least
+// JL_PATH_MAX, and the length of that string (excluding null terminator) to "written"
+// Returns 1 on success or 0 on error, in which case the caller should abort the process.
+#define JL_DEFAULT_STDCXXLIB "libstdc++.so.6"
+static int libstdcxxprobe(char* path, size_t* written) {
+    // On Linux, it can happen that the system as a newer libstdc++ than the one we ship,
+    // which can break loading of some system libraries:
+    // <https://github.com/JuliaLang/julia/issues/34276>.
+    int ret;
+    pid_t pid;
+    int fork_pipe[2];
+    void *handle;
+    char *dlerr;
+
+    // Create the pipe and child process.
+    ret = pipe(fork_pipe);
+    if (ret == -1) return 0;
+    pid = fork();
+    if (pid == -1) return 0;
+
+    if (pid == (pid_t) 0) { // Child process.
+        ret = close(fork_pipe[0]);
+        if (ret == -1) exit(1);
+
+        // Open the first available libstdc++.so
+        handle = dlopen("libstdc++.so", RTLD_LAZY);
+        if (!handle) exit(0);
+
+        // See if the version is compatible
+        dlerr = dlerror(); // clear out dlerror
+        void* sym = dlsym(handle, "GLIBCXX_3.4.29");
+        dlerr = dlerror();
+        if (dlerr) {
+            // We can't use the library that was found, use the default one instead.
+            ret = write(fork_pipe[1], JL_DEFAULT_STDCXXLIB, strlen(JL_DEFAULT_STDCXXLIB));
+            if (ret == -1) exit(1);
+            exit(0);
+        }
+        
+        // No error means the symbol was found, we can use this library/
+        // The symbol we just got is always NULL. Get another one that actually
+        // has an address so we can call dladdr() with it to get the path to the
+        // shared object it was loaded from. The symbol is arbitrary.
+        Dl_info info;
+        sym = dlsym(handle, "__gxx_personality_v0");
+        if (!sym) exit(1);
+        ret = dladdr(sym, &info);
+        if (ret == 0) exit(1);
+        ret = write(fork_pipe[1], info.dli_fname, strlen(info.dli_fname));
+        if (ret == -1) exit(1);
+        exit(0);
+    }
+    else { // Parent process.
+        close(fork_pipe[1]);
+        // Read the absolute path to the lib from the child process.
+        ssize_t have_read = read(fork_pipe[0], path, JL_PATH_MAX);
```

Comment:
this should loop until it returns 0, or returns -1 and errno is not EAGAIN

Reply:
Should `JL_PATH_MAX` be changed to actually be the `PATH_MAX` for the system (from `linux/limits.h`)? I brought it up with Mosè and Elliot, but we decided to not to immediately expand the scope of what we were doing.

Reply:
Also `PATH_MAX` is a lie. There is no maximum limit for the length of a path. It's one of these dark shameful corners of POSIX that everyone hopes will go away if we just all pretend it's not there, but sadly it won't...